### PR TITLE
Core: Introduce stm32mp remoteproc PTA and driver

### PR DESCRIPTION
**Pull request by arnopo**
This series introduces the stm32mp remoteproc PTA and the stm32_remoteproc driver , which enable the authentication of signed images for the stm32mp157 platform.

1) The remote proc TA implements platform-specific requirements associated to the remoteproc PTA API introduced in  https://github.com/OP-TEE/optee_os/pull/6300. The API commit is also part of this PR for build purposes, and should be removed if https://github.com/OP-TEE/optee_os/pull/6300 is merged.
 
2) The stm32mp_remoteproc driver is responsible for relying on the device tree, which is very similar to the Linux one. Note that the remoteproc node is a sub-node of the AHB simple bus. The AHB bus defines the dma-range property that specifoes memory translation between the Cortex-M4 mapping and the Cortex-A7 mapping. This is necessary to find the address translation between the address found in the ELF file and the address to load the segment. For instance, the RETRAM address is 0 from the Cortex-M4 perspective and 0x38000000 from the Cortex-A7 perspective.

For more details a presentation is available here: [LHR23-214 OP-TEE to extend the root of trust to the coprocessor](https://resources.linaro.org/en/resource/6c5bGvZwUAjX56fvxthxds)

Related pull request:

- https://github.com/OP-TEE/optee_os/pull/6027

- https://github.com/OP-TEE/optee_os/pull/6300

- Code associated to the Linux Kernel is proposed as RFC for upstream:

    https://lore.kernel.org/lkml/20230523091350.292221-1-arnaud.pouliquen@foss.st.com/

### CODE REVIEW

**Review by etienne-lms — COMMENTED** [review]
Commit "drivers: Add stm32mp1 remoteproc driver" should define a default value for `CFG_DRIVERS_REMOTEPROC` and provide an inline description comment.

Ditto for commit "pta: stm32mp: add new remoteproc pta" regarding `CFG_RPROC_PTA`.

My main concern here is detailed in my review comment on `stm32_rproc_get()` function.

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+ * @firmware_id:	identifies the remoteproc firmware to load
+ */
+struct stm32_rproc_compat_data {
+	uint32_t firmware_id;
+};
+
+static SLIST_HEAD(, stm32_rproc_instance) rproc_list =
+		SLIST_HEAD_INITIALIZER(rproc_list);
+
+void *stm32_rproc_get(uint32_t firmware_id)
```
[the remark this answers is not in the document]
Comment (arnopo) [reply core/drivers/remoteproc/stm32_remoteproc.c:64]: If I understand correctly @etienne-lms, the question is whether the UUID should guarantee the uniqueness of the service (represented by the PTA API) or the implementation of the PTA itself.
If it guarantees the service, then PTA can be platform dependent. 
If it guarantees the implementation, then the PTA should be generic and we may need to introduce another layer to address specific details that may be too high-level for a driver. 

For instance we could have  to support:
-  management of  a proprietary  image format,
-  specific management of a key to authenticate or decrypt,
-  management of metadata associated to a firmware that could be provided by TLV ,
-  core dump on crash

File: core/pta/stm32mp/remoteproc_pta.c [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+#define PTA_NAME	"remoteproc.pta"
+
+/*
+ * UUID of the remoteproc Trusted application authorized to communicate with
+ * the remoteproc pseudo TA
+ */
+
+#define TA_REMOTEPROC_UUID \
+	{ 0x80a4c275, 0x0a47, 0x4905, \
+		{ 0x82, 0x85, 0x14, 0x86, 0xa9, 0x77, 0x1a, 0x08} }
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/pta/stm32mp/remoteproc_pta.c:29]: You're right, TA's API header file are not exported.
Keep as-is.

File: core/drivers/remoteproc/stm32_remoteproc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jforissier) [reply core/drivers/remoteproc/stm32_remoteproc.c]: I would say the current implementation LGTM. I believe we should consider the remoteproc PTA UUID identifies a service that the remoteproc TA relies on, and that the PTA implementation can be platform specific. I don't see too much code re-use being made possible by introducing another abstraction layer. IOW the PTA interface is probably the right level for separation between generic code (remoteproc TA) and platform-specific code (remoteproc PTA and driver).

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 3]
```diff
@@ -174,14 +174,14 @@ TEE_Result stm32_rproc_map(uint32_t firmware_id, paddr_t pa, size_t size,
...
 }
 
-TEE_Result stm32_rproc_unmap(uint32_t firmware_id, paddr_t pa, size_t size)
+TEE_Result stm32_rproc_unmap(uint32_t rproc_id, vaddr_t *va, size_t size)
 {
-	struct stm32_rproc_instance *rproc = stm32_rproc_get(firmware_id);
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
 	struct stm32_rproc_mem *mems = NULL;
-	vaddr_t *va = NULL;
+	paddr_t pa  = virt_to_phys(va);
```
Comment (etienne-lms) [inline core/drivers/remoteproc/stm32_remoteproc.c:181]: could you replace the 2 space chars with a single one?

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 4]
```diff
@@ -174,14 +174,14 @@ TEE_Result stm32_rproc_map(uint32_t firmware_id, paddr_t pa, size_t size,
 	return TEE_ERROR_ACCESS_DENIED;
 }
 
-TEE_Result stm32_rproc_unmap(uint32_t firmware_id, paddr_t pa, size_t size)
+TEE_Result stm32_rproc_unmap(uint32_t rproc_id, vaddr_t *va, size_t size)
```
Comment (etienne-lms) [inline core/drivers/remoteproc/stm32_remoteproc.c:177]: this passes a va or a pointer to a va?
I think it should be either `void *va` or `vaddr_t va` here.

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 5]
```diff
@@ -174,14 +174,14 @@ TEE_Result stm32_rproc_map(uint32_t firmware_id, paddr_t pa, size_t size,
...
 {
-	struct stm32_rproc_instance *rproc = stm32_rproc_get(firmware_id);
+	struct stm32_rproc_instance *rproc = stm32_rproc_get(rproc_id);
 	struct stm32_rproc_mem *mems = NULL;
-	vaddr_t *va = NULL;
+	paddr_t pa  = virt_to_phys(va);
 	unsigned int i = 0;
 
-	if (!rproc)
+	if (!rproc || !va)
```
Comment (etienne-lms) [inline core/drivers/remoteproc/stm32_remoteproc.c:184]: `!va` or `!pa`  (note: pa will be 0 if va is 0)

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 6]
```diff
@@ -0,0 +1,389 @@
...
+			goto err;
+		}
+
+		DMSG("register region %#"PRIxPA" size %#zx",
+		     regions[i].addr, regions[i].size);
+
+		rproc->n_regions++;
+	}
+
+	rproc->regions = regions;
```
[the remark this answers is not in the document]
Comment (etienne-lms) [reply core/drivers/remoteproc/stm32_remoteproc.c:310]: indeed, `n_regions` does not change. my bad. discard my comment.

File: core/pta/stm32mp/remoteproc_pta.c [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+static TEE_Result
+	rproc_pta_open_session(uint32_t pt, TEE_Param params[TEE_NUM_PARAMS],
+			       void **sess_ctx __unused)
+{
+	struct ts_session *s = ts_get_calling_session();
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_NONE);
+	struct ts_ctx *ctx;
```
Comment (etienne-lms) [inline core/pta/stm32mp/remoteproc_pta.c:333]: = NULL

**Review by etienne-lms — COMMENTED** [review]
I have few comments on the fixup commits. I'm fine if you squash and rebase once addressed.

**Review by arnopo — COMMENTED** [review]
As requested by @etienne-lms, I have updated the PR based on comment, squashed the fixups  and rebased the PR.
On top of that I  have added a new fixup that adds additional  fixes.

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+
+	if ((len % (sizeof(uint32_t) * 3)))
+		return TEE_ERROR_GENERIC;
+
+	nranges = len / sizeof(uint32_t);
+
+	for (i = 0; i < nranges; i++) {
+		uint32_t da = fdt32_to_cpu(list[i++]);
+		uint32_t pa = fdt32_to_cpu(list[i++]);
+		uint32_t size = fdt32_to_cpu(list[i]);
```
Comment (etienne-lms) [inline core/drivers/remoteproc/stm32_remoteproc.c:242]: nitpicking: here I would prefer:
```c
	for (i = 0; i < nranges; i += 3) {
		uint32_t da = fdt32_to_cpu(list[i]);
		uint32_t pa = fdt32_to_cpu(list[i + 1]);
		uint32_t size = fdt32_to_cpu(list[i + 2]);
		(...)

File: core/pta/stm32mp/remoteproc_pta.c [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+/*
+ * Pseudo Trusted Application entry points
+ */
+static TEE_Result
+	rproc_pta_open_session(uint32_t pt, TEE_Param params[TEE_NUM_PARAMS],
+			       void **sess_ctx __unused)
```
Comment (etienne-lms) [inline core/pta/stm32mp/remoteproc_pta.c:326]: prefer:
```c
static TEE_Result rproc_pta_open_session(uint32_t pt,
					 TEE_Param params[TEE_NUM_PARAMS],
					 void **sess_ctx __unused)

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits 
"drivers: Add stm32mp1 remoteproc driver" and
"pta: stm32mp: add new remoteproc pta" with a comment in each addressed.

**Comment by arnopo** [comment]
> `Reviewed-by: Etienne Carriere <etienne.carriere@foss.st.com>` for commits "drivers: Add stm32mp1 remoteproc driver" and "pta: stm32mp: add new remoteproc pta" with a comment in each addressed.

Done, thanks @etienne-lms for the reviews!

File: mk/config.mk [hunk 10]
```diff
@@ -1132,3 +1132,10 @@ $(eval $(call cfg-depends-all,CFG_CRYPTO_HW_PBKDF2,CFG_CRYPTO_PBKDF2))
 CFG_HALT_CORES_ON_PANIC ?= n
 CFG_HALT_CORES_ON_PANIC_SGI ?= 15
 $(eval $(call cfg-depends-all,CFG_HALT_CORES_ON_PANIC,CFG_GIC))
+
+# CFG_DRIVERS_REMOTEPROC, when enabled, embeds support for remote processor
+# management including generic DT bindings for the configuration.
+CFG_DRIVERS_REMOTEPROC ?= n
+
+# CFG_RPROC_PTA, when enabled, embeds remote processor management PTA service.
+CFG_RPROC_PTA ?= n
```
Comment (jforissier) [inline mk/config.mk:1141]: Would you mind `s/CFG_RPROC_PTA/CFG_REMOTEPROC_PTA/` for consistency with `CFG_DRIVERS_REMOTEPROC` (and better "grepability", too)?

File: core/drivers/remoteproc/stm32_remoteproc.c [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+	 * the hardware.
+	 */
+	res = rstctrl_deassert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+	if (!res)
+		res = rstctrl_assert_to(rproc->hold_boot, TIMEOUT_US_1MS);
+
+	return res;
+}
+
+static TEE_Result __stm32_rproc_stop(struct stm32_rproc_instance *rproc)
```
Comment (jforissier) [inline core/drivers/remoteproc/stm32_remoteproc.c:96]: Leading underscores/double underscores are not really recommended, why not simply `rproc_stop()`? It is a static function the name can remain very short and generic with no risk of collision.

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>`for commits
"plat-stm32mp1: Add the remoteproc TA in early-TA list" and
"dts: stm32: update m4_rproc to support the remoteproc OPTEE framework".

**Review by jforissier — COMMENTED** [review]
Minor comments on "drivers: Add stm32mp1 remoteproc driver".

s/coherence/consistency/ in the commit description. Other comments below.

File: core/arch/arm/plat-stm32mp1/conf.mk [hunk 12]
```diff
@@ -151,6 +151,7 @@ CFG_CORE_ASLR ?= n
 
 CFG_STM32MP_REMOTEPROC ?= n
 CFG_DRIVERS_REMOTEPROC ?= $(CFG_STM32MP_REMOTEPROC)
+CFG_RPROC_PTA ?= $(CFG_STM32MP_REMOTEPROC)
```
Comment (jforissier) [inline core/arch/arm/plat-stm32mp1/conf.mk:154]: As already mentioned I would prefer `CFG_REMOTEPROC_PTA`

**Review by jforissier — COMMENTED** [review]
For commit " pta: stm32mp: add new remoteproc pta"

s/pta/PTA/ in the suject (the `pta:` prefix is OK in lowercase).

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "plat-stm32mp1: Add the remoteproc TA in early-TA list":

s/early-TA/early TA/

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/drivers/remoteproc/stm32_remoteproc.c (hunk 11 above)
Comment (arnopo) [reply core/drivers/remoteproc/stm32_remoteproc.c]: Done

File: core/arch/arm/plat-stm32mp1/conf.mk (hunk 12 above)
Comment (arnopo) [reply core/arch/arm/plat-stm32mp1/conf.mk]: Done

File: mk/config.mk (hunk 10 above)
Comment (arnopo) [reply mk/config.mk]: yes much better  I also move the declaration in the patch that introduce the remoteproc PTA

**Review by arnopo — COMMENTED** [review]
@jforissier : I have take, into account your comments. In addition I have re-dispatched the CFG-REMOTEPROC-PTA between commits. As consequence, I don't add your tag and @etienne-lms one in commits.
Please, could you confirm that the fixups are ok for you then i will squash them add the tags in commits messages.

**Comment by jforissier** [comment]
@arnopo yes the fixups look good, thanks. When squashing you may also add:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

... to "drivers: Add stm32mp1 remoteproc driver".

**Comment by arnopo** [comment]
+1 update to fix a typo in commit adding the acked-by

**Pull request merged**